pkg/azuredx/adxauth/adxcredentials/builder.go

```diff
@@ -0,0 +1,171 @@
+	authType, err := maputil.GetString(credentialsObj, "authType")
+	if err != nil {
+		return nil, err
+	}
+
+	switch authType {
+	case AzureAuthClientSecretObo:
+		if err := ensureOnBehalfOfSupported(data); err != nil {
+			return nil, err
+		}
+
+		cloud, err := maputil.GetString(credentialsObj, "azureCloud")
+		if err != nil {
+			return nil, err
+		}
+		tenantId, err := maputil.GetString(credentialsObj, "tenantId")
+		if err != nil {
+			return nil, err
+		}
+		clientId, err := maputil.GetString(credentialsObj, "clientId")
+		if err != nil {
+			return nil, err
+		}
+		clientSecret := secureData["azureClientSecret"]
+
+		credentials := &AzureClientSecretOboCredentials{
+			ClientSecretCredentials: azcredentials.AzureClientSecretCredentials{
+				AzureCloud:   cloud,
+				TenantId:     tenantId,
+				ClientId:     clientId,
+				ClientSecret: clientSecret,
+			},
+		}
+		return credentials, nil
+
+	default:
+		return nil, nil
+	}
+}
+
+func getFromLegacy(data map[string]interface{}, secureData map[string]string) (azcredentials.AzureCredentials, error) {
+	legacyCloud, err := maputil.GetStringOptional(data, "azureCloud")
+	if err != nil {
+		return nil, err
+	}
+	cloud, err := normalizeAzureCloud(legacyCloud)
+	if err != nil {
+		return nil, fmt.Errorf("invalid Azure credentials: %w", err)
+	}
+	tenantId, err := maputil.GetStringOptional(data, "tenantId")
+	if err != nil {
+		return nil, err
+	}
+	clientId, err := maputil.GetStringOptional(data, "clientId")
+	if err != nil {
+		return nil, err
+	}
+	clientSecret := secureData["clientSecret"]
+
+	// If any of the required fields are not set then credentials are not configured
+	if tenantId == "" || clientId == "" || clientSecret == "" {
+		return nil, nil
+	}
+
+	clientSecretCredentials := azcredentials.AzureClientSecretCredentials{
+		AzureCloud:   cloud,
+		TenantId:     tenantId,
+		ClientId:     clientId,
+		ClientSecret: clientSecret,
+	}
+
+	onBehalfOf, err := maputil.GetBoolOptional(data, "onBehalfOf")
+	if err != nil {
+		return nil, err
+	}
+
+	var credentials azcredentials.AzureCredentials
+
+	if onBehalfOf {
+		if err := ensureOnBehalfOfSupported(data); err != nil {
+			return nil, err
+		}
+
+		credentials = &AzureClientSecretOboCredentials{
+			ClientSecretCredentials: clientSecretCredentials,
+		}
+	} else {
+		credentials = &clientSecretCredentials
+	}
+
+	return credentials, nil
+}
+
+func ensureOnBehalfOfSupported(data map[string]interface{}) error {
+	oauthPassThru, err := maputil.GetBoolOptional(data, "oauthPassThru")
+	if err != nil {
+		return err
+	} else if !oauthPassThru {
+		return errors.New("oauthPassThru should be enabled for On-Behalf-Of authentication")
+	} else {
+		return nil
+	}
+}
+
+// Azure cloud names used by the Azure Data Explorer datasource
+const (
+	azureMonitorPublic       = "azuremonitor"
+	azureMonitorChina        = "chinaazuremonitor"
+	azureMonitorUSGovernment = "govazuremonitor"
+)
```

Comment:
They are ADX specific unfortunately.
Builder will have to stay to support legacy format and this is part of the legacy format.

Reply:
even if these names are only used in ADX, it would be useful to move these aliases to the SDK so we can just use the method [`NormalizeAzureCloud`](https://github.com/grafana/grafana-azure-sdk-go/blob/main/azsettings/clouds.go#L11) and have defined there all the possible names.

Reply:
The `azsettings.NormalizeAzureCloud` func was introduced for parsing of INI config, to make the logic more tolerant to user input.
It wasn't intended for this purpose as you suggest.

I would not add anything to `azsettings.NormalizeAzureCloud` because I would not want user to be able to do this:
```ini
[azure]
cloud = azuremonitor
```

Reply:
okay, can you make a comment here then, explaining why these have not been migrated to the sdk?

Reply:
Renamed the method to `resolveLegacyCloudName` and mentioned that the cloud names are legacy.